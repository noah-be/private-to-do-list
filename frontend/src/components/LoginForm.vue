<template>
  <form id="login-form" @submit.prevent="submitLogin">
    <RegisterBox v-if="showregisterBox" @close="showregisterBox = false" />

    <h2>Login</h2>
    <label for="login-username">Username</label>
    <input
      type="text"
      id="login-username"
      v-model="username"
      placeholder="Username"
      autocomplete="username"
    />

    <label for="login-password">Password</label>
    <input
      type="password"
      id="login-password"
      v-model="password"
      placeholder="Password"
      autocomplete="current-password"
    />

    <button id="login-btn" type="submit">Log In</button>
    <hr />
    <button
      id="create-new-account-btn"
      class="create-new-account-btn"
      @click.prevent="createNewAccount"
    >
      Create new account
    </button>
  </form>
</template>

<script>
import RegisterBox from "@/components/RegisterBox.vue";

export default {
  components: {
    RegisterBox,
  },
  data() {
    return {
      username: "",
      password: "",
      showregisterBox: false,
    };
  },
  methods: {
    async submitLogin() {
      try {
        const response = await fetch("/api/auth/login", {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            username: this.username,
            password: this.password,
          }),
        });

        if (!response.ok) {
          throw new Error("Login failed");
        }

        const data = await response.json();
        console.debug("Success:", data);
        window.location.href = "/tasks";
      } catch (error) {
        console.error("Error:", error);
        alert("Login failed. Please try again.");
      }
    },
    createNewAccount() {
      this.showregisterBox = true;
    },
  },
};
</script>
