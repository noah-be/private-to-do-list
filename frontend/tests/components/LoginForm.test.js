import { describe, it, expect, vi, beforeAll } from "vitest";
import { mount } from "@vue/test-utils";
import LoginForm from "@/components/LoginForm.vue";
import { createRouter, createWebHistory } from "vue-router";

const DummyComponent = { template: "<div>Dummy</div>" }; // Dummy Component for router setup

// Mock router setup
const mockRoutes = [
  { path: "/", name: "home", component: DummyComponent },
  { path: "/tasks", name: "tasks", component: DummyComponent },
];
const mockRouter = createRouter({
  history: createWebHistory(),
  routes: mockRoutes,
});

beforeAll(() => {
  global.alert = vi.fn(); // Mock window.alert
  global.fetch = vi.fn(); // Mock fetch API
});

describe("LoginForm.vue", () => {
  it("matches the snapshot", () => {
    const wrapper = mount(LoginForm, {
      global: {
        plugins: [mockRouter],
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("shows the RegisterBox when the create new account button is clicked", async () => {
    const wrapper = mount(LoginForm, {
      global: {
        plugins: [mockRouter],
      },
    });

    await wrapper.find("#create-new-account-btn").trigger("click");
    expect(wrapper.findComponent({ name: "RegisterBox" }).exists()).toBe(true);
  });

  it("sends login request with correct data", async () => {
    global.fetch.mockResolvedValueOnce({
      ok: true,
      json: () => Promise.resolve({ token: "123", redirectUrl: "/tasks" }),
    });

    const wrapper = mount(LoginForm, {
      global: {
        plugins: [mockRouter],
      },
    });

    await wrapper.find("#login-username").setValue("testuser");
    await wrapper.find("#login-password").setValue("testpass");
    await wrapper.find("#login-form").trigger("submit.prevent");

    expect(global.fetch).toHaveBeenCalledWith("/api/auth/login", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ username: "testuser", password: "testpass" }),
    });
  });

  it("handles login errors correctly", async () => {
    global.fetch.mockResolvedValueOnce({ ok: false });

    const wrapper = mount(LoginForm, {
      global: {
        plugins: [mockRouter],
      },
    });

    await wrapper.find("#login-username").setValue("testuser");
    await wrapper.find("#login-password").setValue("wrongpass");
    await wrapper.find("#login-form").trigger("submit.prevent");

    expect(global.fetch).toHaveBeenCalled();
    expect(global.alert).toHaveBeenCalledWith(
      "Login failed. Please try again.",
    );
  });
});
