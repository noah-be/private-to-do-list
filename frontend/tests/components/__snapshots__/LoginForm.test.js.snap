// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm.vue > matches the snapshot 1`] = `
"<form id="login-form">
  <!--v-if-->
  <h2>Login</h2><label for="login-username">Username</label><input type="text" id="login-username" placeholder="Username" autocomplete="username"><label for="login-password">Password</label><input type="password" id="login-password" placeholder="Password" autocomplete="current-password"><button id="login-btn" type="submit">Log In</button>
  <hr><button id="create-new-account-btn" class="create-new-account-btn"> Create new account </button>
</form>"
`;
